First draft for simulation and workflow schemas

mostaphaRoudsari opened the pull request:
I need to add sample files and tests for schemas before we merge this in but I thought I should send this PR to get an initial feedback.

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+    name = sa.Column(sa.String)
+    type = sa.Column(sa.Enum(TypeEnum), nullable=False)
+    kind = sa.Column(sa.Enum(KindEnum))
+    simulation = relationship('Workflow', back_populates='simulations')
```
Is this meant to be `workflow`?

mostaphaRoudsari replied:
Yes. probably. I will check.

mostaphaRoudsari replied:
I did double check this and it seems 'Workflow' is correct and it is referring to `Workflow` class.

https://docs.sqlalchemy.org/en/latest/orm/basic_relationships.html#one-to-many

AntoineDao replied:
Oh sorry I meant the variable. Should it be this:
```python
    workflow = relationship('Workflow', back_populates='simulations')
```

AntoineDao commented on app/models/radiance/sky/climatebased.py:
```diff
@@ -0,0 +1,25 @@
+    )
+    location: Location
+    datetime: DateTime
+    direct_irradiance: float = Schema(
```
Are we sending an entire 8760 x 2 point array in one go? This is a pretty massive payload and I reckon we want to manage it somewhere else or allow some sort of pagination no?

mostaphaRoudsari replied:
It's only 2 values (for a single hour) since this sky is for a single point-in-time study. Once we get to annual studies I think we can start thinking about another service for sky. For point-in-time studies it will be very light.

AntoineDao replied:
Ah right my bad, got confused as it was put under `climatebased`. We'll cross the yearly bridge when we get there 😄

AntoineDao commented on app/models/samples/simulation.py:
```diff
@@ -0,0 +1,68 @@
+    "outputs":
+      [
+        {
+          "verb": "post",
```
Why do you post a specific endpoint to get data back? POST verb is for creating data usually.

mostaphaRoudsari replied:
This can change but the reasoning behind it is that most of outputs end up creating new resources. I'm personally fine with using the one that makes the most sense from user point of view unless we get someone to be the HTTP 👮/👮‍♀️ here and enforce one over the other.

mostaphaRoudsari replied:
Ah! I remember why I didn't use `GET` more than the fact that they will create new issues. Technically a `GET` request should not have a body in the request and for these calls we need to send a body.

AntoineDao commented on app/models/samples/simulation.py:
```diff
@@ -0,0 +1,68 @@
+      [
+        {
+          "description": "Start the simulation.",
+          "verb": "put",
```
This feels like a POST verb but can do PUT if you're keen on it.

mostaphaRoudsari replied:
Sounds good to me. Will change it.

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+    simulation_id = sa.Column(sa.String, sa.ForeignKey('simulation.id'))
+    resource_id = sa.Column(sa.String, nullable=False)
+    resource_type = sa.Column(sa.Enum(ResourceEnum), nullable=False)
+    resource = relationship('Simulation', back_populates='resources')
```
Where is the `resources` table?

mostaphaRoudsari replied:
Hmmm. It refers to `Simulation.resources`. See line 72. Or did you mean _what_ is the resources table?

AntoineDao replied:
Ah right I get you, so this is a simulation then:
```python
simulation = relationship('Simulation', back_populates='resources')
```

AntoineDao commented on app/models/radiance/simulation/daylightfactor.py:
```diff
@@ -0,0 +1,29 @@
+    radiance_parameters: RadianceParameters = None
+
+
+class DaylightFactorGridBasedOut(GridBasedOut):
```
Do we not want to see what the radiance parameters are when we GET this resource?

mostaphaRoudsari replied:
I would say so. Let's get it in.

AntoineDao commented on app/api/simulation.py:
```diff
@@ -0,0 +1,283 @@
+    summary='Get Simulations',
+    status_code=HTTP_200_OK,
+    response_description='Retrieved',
+    response_model=List[Union[DaylightFactorGridBasedOut, PointInTimeGridBasedOut]],
```
This works for now but we'll probably have to come up with a broader definition for an output response model once we have more than 2 different simulations available.

mostaphaRoudsari replied:
Agreed! We should probably collect all `resources` for each simulation under a separate field and collect all the resources so it can handle different simulation types. Let's address this once the time comes.

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+    event_type = sa.Column(sa.Enum(EventEnum), nullable=False)
+    timestamp = sa.Column(sa.DateTime, nullable=False,
+        default=datetime.datetime.utcnow())
+    event = relationship('Simulation', back_populates='events')
```
```
simulation = relationship('Simulation', back_populates='events')
```

mostaphaRoudsari replied:
done!

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+    id = sa.Column(sa.String, primary_key=True, default=str(uuid.uuid4()))
+    simulation_id = sa.Column(sa.String, sa.ForeignKey('simulation.id'))
+    dependent_id = sa.Column(sa.String, nullable=False)
+    dependent = relationship('Simulation', back_populates='dependents')
```
```python
 simulation = relationship('Simulation', back_populates='dependents')
```

mostaphaRoudsari replied:
👍

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+        cascade='all, delete, delete-orphan'
+    )
+    events = relationship(
+        'SimulationEvent', back_populates='event',
```
```python
    events = relationship(
        'SimulationEvent', back_populates='simulation',
        cascade='all, delete, delete-orphan'
    )
```

mostaphaRoudsari replied:
Done!

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+        'SimulationDependent', back_populates='dependent',
+        cascade='all, delete, delete-orphan'
+    )
+    resources = relationship(
```
```python
    resources = relationship(
        'SimulationResource', back_populates='simulation',
        cascade='all, delete, delete-orphan'
    )
```

mostaphaRoudsari replied:
Fixed!

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+    type = sa.Column(sa.Enum(TypeEnum), nullable=False)
+    kind = sa.Column(sa.Enum(KindEnum))
+    simulation = relationship('Workflow', back_populates='simulations')
+    dependents = relationship(
```
```python
    dependents = relationship(
        'SimulationDependent', back_populates='simulation',
        cascade='all, delete, delete-orphan'
    )
```

mostaphaRoudsari replied:
Thanks! Fixed this one too.

AntoineDao commented on app/models/database.py:
```diff
@@ -0,0 +1,110 @@
+                           default=datetime.datetime.utcnow())
+    name = sa.Column(sa.String)
+    simulations = relationship(
+        "Simulation", back_populates='simulation',
```
```python
    simulations = relationship(
        'Simulation', back_populates='workflow',
        cascade='all, delete, delete-orphan'
    )
```

mostaphaRoudsari replied:
fixed!

AntoineDao commented on app/models/samples/simulation.py:
```diff
@@ -0,0 +1,68 @@
+    "outputs":
+      [
+        {
+          "verb": "put",
```
Not sure I understand what these outputs are. Are they endpoints for the user to call to get that data? If so shouldn't it be get?

mostaphaRoudsari replied:
They are endpoints to be called but the workflow needs two steps:

1. Request an output with a `POST` request. The request will return a new link to access the output itself.

2. Make a GET call to the new link.

This might look like an extra step but we will need this for most of outputs. Average, more than x, less than x, daylight autonomy, useful daylight illuminance, and so on. Also see here: https://github.com/pollination/api/pull/24/files#diff-2c036c79e98494ce3df21b5e88dbbeebR298

AntoineDao replied:
Ok let's leave this as is for now. I reckon there's a chance we'll either just run all the basic post-process steps at the end of a job as it's minimal compute compared to the actual simulation or we might turn these steps into smaller `simulation` objects that carry out their own post-processing.

AntoineDao commented on app/models/samples/simulation.py:
```diff
@@ -0,0 +1,68 @@
+
+simulation =   {
+    "name": "market_place_comfort_study",
+    "id": "123e4567-e89b-12d3-a456-426655440000",
```
What's this combined `uuid`-`long-number`?

mostaphaRoudsari replied:
No comments!  🤒

AntoineDao replied:
My question is: is it meant to be a uuid? Or is there some pattern I'm not understanding here?

mostaphaRoudsari replied:
It is supposed to be a UUID. but it is handmade in this case. I'm fixing this and adding tests for samples.

AntoineDao replied:
Right I get you 😄

mostaphaRoudsari commented:
@AntoineDao thank you for reviewing the code and suggestions.

1. Removed! I will send a separate PR for that.
2. Fixed. Thank you for the suggestions.
